[FIX] Scheduling customization adjustments

eva-vashkevich opened the pull request:
<!-- This template is for Devs to give QA details before moving the issue To-Test -->
### Summary
Fixes #13615 #13611 #13630
<!-- Define findings related to the feature or bug issue. -->
- Hid  'Cluster agent' tab for local cluster
- Adjusted Scheduling customization behavior when feature is disabled

### Occurred changes and/or fixed issues
<!-- Include information of the changes, including collateral areas which have been affected by this PR as requirement or for convenience. -->

### Technical notes summary
<!-- Outline technical changes which may pass unobserved or may help to understand the process of solving the issue -->

### Areas or cases that should be tested
<!-- Areas that should be tested can include Airgap checks, Rancher upgrades, K8s upgrade, etc. -->
<!-- Which browser did you use for local testing? The reviewer should test with a different browser. -->
<!-- Add missing steps or rewrite them if have been missed or to complement existing information. This should define a clear way to reproduce it and not an approximation. -->
1. Enable 'cluster-agent-scheduling-customization' feature flag
2. Cluster management -> Edit Local cluster -> 'Cluster agent' tab should be hidden
3. When editing local cluster, 'clusterAgentDeploymentCustomization.schedulingCustomization' field should not be present in the request
3. Provision a new RKE2 cluster
4. Have Scheduling Customization enabled
5. Disable 'cluster-agent-scheduling-customization' feature flag 
6. Go back to the cluster you just created
7. Scheduling Customization  should be enabled, but 'Apply global settings for Priority Class and Pod Disruption Budget' should be hidden

### Areas which could experience regressions
<!-- Create a detailed list of areas to be analyzed which may be affected by the changes, which would require a prior research to avoid regressions. -->
Imported cluster behavior can be affected

### Screenshot/Video
<!-- Attach screenshot or video of the changes and eventual comparison if you find it necessary -->

### Checklist
- [x] The PR is linked to an issue and the linked issue has a Milestone, or no issue is needed
- [x] The PR has a Milestone <!-- The milestone should automatically be assigned if the linked issue has one, but does not always happen (incorrectly linked, issue has no milestone, etc) -->
- [x] The PR template has been filled out
- [x] The PR has been self reviewed <!-- There are no TODOs, no incorrect files in the PR, all the required files are there, no commented out code, etc-->
- [x] The PR has a reviewer assigned
- [x] The PR has automated tests or clear instructions for manual tests and the linked issue has appropriate QA labels, or tests are not needed
- [x] The PR has reviewed with UX and tested in light and dark mode, or there are no UX changes

richard-cox commented:
@HarrisonWAffel The RFC has the text

```
Disabling this feature flag will not impact clusters previously provisioned using this feature, but will prevent further reconfiguration of these fields.
```

In the description for the FF we have 

```
Enables the automatic deployment of Pod Disruption Budgets and Priority Classes when deploying the cattle-cluster-agent. Disabling this feature will not impact existing clusters.
```

Is the last bit not true then, what happens to a cluster that has the feature configured and in use but then the FF is disabled? Does nothing happen... until the user edits the cluster and the feature is then removed(and resources removed)

```
Disabling this feature will not impact immediately impact existing clusters, but any subsequent changes to the cluster will result in the feature being removed
```

Or would it be the case that the feature continues to be used (and they can change the config), the user would have to disable at the cluster level?

richard-cox reviewed (commented):
Will leave the full review until after Harrison response to the PR comment. Noticed some things we should do after 2.11At some point though 

- `initSchedulingCustomization` is duplicated in two files
-  `export const SCHEDULING_CUSTOMIZATION = create('cluster-agent-scheduling-customization', false);` should be updated to use `SCHEDULING_CUSTOMIZATION`
- `initSchedulingCustomization` should catch exceptions for `JSON.parse`
- initSchedulingCustomization and setSchedulingCustomization feel like they should be in AgentConfiguration (to remove complexity in the rke2 component)
- settingMissmatch contains some super line `if` conditions that need breaking up

HarrisonWAffel commented:
@richard-cox 

> Is the last bit not true then, what happens to a cluster that has the feature configured and in use but then the FF is disabled? Does nothing happen... until the user edits the cluster and the feature is then removed

After disabling the FF nothing happens to existing downstream clusters until the user removes the related fields spec. Updates to other fields should not automatically clear the fields related to this feature. 

> Or would it be the case that the feature continues to be used (and they can change the config), the user would have to disable at the cluster level?

That's correct, except they cannot change the config related to the PC and PDB other than outright removing it.

richard-cox commented on shell/components/form/SchedulingCustomization.vue:
```diff
@@ -42,7 +42,7 @@ export default {
       const pcMismatch = !!this.value?.priorityClass && (this.value?.priorityClass.value !== this.defaultPC.value || this.value?.priorityClass.preemptionPolicy !== this.defaultPC.preemptionPolicy);
 
-      return this.isEdit && (pdbMismatch || pcMismatch);
+      return this.feature && this.isEdit && (pdbMismatch || pcMismatch);
```
logically this would be better as a condition on the if, rather than implying there's a settings Missmatch

richard-cox reviewed (changes requested):
Haven't given this much of a functional test (hoping QA of issues can handle that).

There's one minor thing that we should probably do (comment inline)

I pick up on a number of things we need to do post 2.11.0 https://github.com/rancher/dashboard/pull/13617#pullrequestreview-2664187298

eva-vashkevich commented:
Hi @richard-cox I've addressed some of the comments you listed and  issue #13630. I think the rest we can look into later on.
- rancher.io annotations should be classes as system labels/annotations (to hide them from users by default
- initSchedulingCustomization should catch exceptions for JSON.parse
- settingMissmatch contains some super line if conditions that need breaking up

richard-cox commented on pkg/imported/components/CruImported.vue:
```diff
@@ -241,7 +246,11 @@ export default defineComponent({
       if (this.isEdit) {
-        return await this.normanCluster.save();
+        if (this.needsReplace) {
+          return await this.normanCluster._save({ replace: true });
```
just `save` can be used
- neither cluster nor norman model's override `save`, so the default resource class one is used
- the default resource class `save` passes it's arguments through to `_save`

richard-cox replied:
This also means this if block can be simplified to just

```
return await this.normanCluster.save({ replace: this.needsReplace });
```

richard-cox commented on shell/models/cluster.js:
```diff
@@ -14,6 +14,7 @@ export const ANNOTATIONS_CONTAINS_PROTECTED = [
   'k3s.io',
   'kubernetes.io',
   'k3s.io',
+  'rancher.io'
```
i'm trying to find out why the hell i suggested this, as it's not referenced in this PR or RFC. might have context switched and made an error. if the feature doesn't use it, it can be removed

richard-cox reviewed (changes requested):
Two small things, and one UX bug
- FF enabled, create cluster with feature on
- edit cluster and click on `Prevent Rancher cluster agent pod eviction` checkbox
- section immediately disappears. in this scenario it shouldn't affect the visibility

Think we should cap any changes to a day though, anything remaining goes into the issue with https://github.com/rancher/dashboard/pull/13617#pullrequestreview-2664187298 (can you create one?). I'll do the final review and merge tomorrow UK AM

eva-vashkevich commented:
@richard-cox I've opened https://github.com/rancher/dashboard/issues/13665 to follow up. I kind of disagree with some of it, so let's have a chat before I make those changes?
